homeassistant/helpers/collection.py

```diff
@@ -343,11 +349,14 @@ async def _remove_entity(change_set: CollectionChangeSet) -> None:
         )
         if ent_to_remove is not None:
             ent_reg.async_remove(ent_to_remove)
-        else:
+        elif change_set.item_id in entities:
             await entities[change_set.item_id].async_remove(force_remove=True)
-        entities.pop(change_set.item_id)
+        if change_set.item_id in entities:
```

Comment:
Removing the entity registry will remove the entity and so will call the `async_on_remove` callback too. So this is not necessary ?

Reply:
If we don't pop the entity from the entities list, we introduce a race where the entity is scheduled for removal but will still be in the `entities` list until the entity registry event has been handled.
I added a comment about that.